### PR TITLE
[NOSQUASH] Update credits for 5.4.0

### PR DESCRIPTION
**Pull request by sfan5**
point system explanation see https://github.com/minetest/minetest/pull/9593#issue-398677198
I rewrote the script in python because it was awful.

## To do

This PR is a Work in Progress.

- [x] properly format lists in tab_credits.lua like they were before (*help gladly accepted*)

### CODE REVIEW

File: builtin/mainmenu/tab_credits.lua [hunk 1]
```diff
@@ -26,25 +26,35 @@ local core_developers = {
...
+	"Lejo",
+	"Paul Ouellette",
+	"TheTermos",
+	"EvidenceB",
+	"David CARLIER",
+	"Danila Shutov",
+	"ClobberXD",
+	"Jean-Patrick Guerrero",
+	"k.h.lai",
+	"ANAND",
```
Comment (LoneWolfHT) [inline builtin/mainmenu/tab_credits.lua:57]: ANAND == ClobberXD and left Minetest last year. So they won't be active anymore

**Comment by rubenwardy** [comment]
I suggest adding the description from #9593 to the top of the python file, to make it clear what it does

File: builtin/mainmenu/tab_credits.lua [hunk 2]
```diff
@@ -63,27 +73,30 @@ local previous_core_developers = {
...
+	"Desour",
+	"MirceaKitsune",
+	"MoNTE48",
+	"Constantin Wenger",
+	"Jean-Patrick Guerrero",
+	"Ciaran Gultnieks",
+	"Paul Ouellette",
+	"Zughy",
+	"stujones11",
+	"Muhammad Rifqi Priyo Susanto",
```
Comment (rubenwardy) [inline builtin/mainmenu/tab_credits.lua:95]: This is srifqi

File: util/gather_git_credits.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
+import subprocess
+import re
+from collections import defaultdict
+
+codefiles = r"(\.[ch](pp)?|\.lua|\.md|\.cmake|\.java|\.gradle|Makefile|CMakeLists\.txt)$"
+
+# two major versions back, for "Active Contributors"
```
Comment (rubenwardy) [inline util/gather_git_credits.py:8]: minor versions

File: builtin/mainmenu/tab_credits.lua (hunk 1 above)
Comment (SmallJoker) [reply builtin/mainmenu/tab_credits.lua:57]: Yet they actively contributed to 5.4.0, hence an attribution here for ANAND/ClobberXD is technically correct.

File: builtin/mainmenu/tab_credits.lua [hunk 4]
```diff
@@ -63,27 +73,30 @@ local previous_core_developers = {
...
-	"TeTpaAka [Hand overriding, nametag colors]",
-	"Rui [Sound Pitch]",
-	"Duane Robertson <duane@duanerobertson.com> [MGValleys]",
-	"Raymoo [Tool Capabilities]",
-	"Rogier <rogier777@gmail.com> [Fixes]",
-	"Gregory Currie (gregorycu) [optimisation]",
-	"TriBlade9 <triblade9@mail.com> [Audiovisuals]",
-	"T4im [Profiler]",
-	"Jurgen Doser (doserj) <jurgen.doser@gmail.com>",
+	"numzero",
```
Comment (SmallJoker) [inline builtin/mainmenu/tab_credits.lua:80]: I think "previous contributors" should exclude active ones to keep things shorter.

File: util/gather_git_credits.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+revs = "5.2.0..HEAD"
+# all time, for "Previous Contributors"
+#revs = "HEAD"
```
Comment (rubenwardy) [inline util/gather_git_credits.py:11]: I suggest refactoring this into a function, and calling it twice

File: builtin/mainmenu/tab_credits.lua (hunk 4 above)
Comment (sfan5) [reply builtin/mainmenu/tab_credits.lua:80]: Yes I merely forgot that.

File: util/gather_git_credits.py (hunk 5 above)
Comment (sfan5) [reply util/gather_git_credits.py]: It now does the whole job including deduplicating contributors so they only appear once in "active" or "previous"

File: util/gather_git_credits.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+CUTOFF_ACTIVE = 3
+CUTOFF_PREVIOUS = 21
+
+# For a description of the points system see:
+# https://github.com/minetest/minetest/pull/9593#issue-398677198
+
+def load(revs):
+	points = defaultdict(int)
+	p = subprocess.Popen(["git", "log", "--mailmap", "--pretty=format:%h %aN <%aE>", revs],
```
Comment (appgurueu) [inline util/gather_git_credits.py:21]: This only works with old git: https://stackoverflow.com/a/27826831/7185318

Comment (sfan5) [reply util/gather_git_credits.py:21]: The man page lists both `--use-mailmap` and `--mailmap` and this works for me on git 2.30.1 which is as new as it gets.

File: builtin/mainmenu/tab_credits.lua [hunk 7]
```diff
@@ -26,25 +26,34 @@ local core_developers = {
...
+	"Desour",
+	"HybridDog",
+	"Hugues Ross",
+	"ClobberXD",
+	"Vincent Robinson",
+	"luk3yx",
+	"hecks",
+	"LoneWolfHT",
+	"Lejo",
+	"EvidenceB",
```
Comment (rubenwardy) [inline builtin/mainmenu/tab_credits.lua:50]: not a person - but a company. Not sure how best to credit, this results in some inconsistency with pyrollo and kilbith

**Comment by appgurueu** [comment]
Suggestion: Fetch usernames from the GitHub API.

**Comment by EliasFleckenstein03** [comment]
Can you please re-run the script after pulling the latest minetest commits?

File: builtin/mainmenu/tab_credits.lua [hunk 8]
```diff
@@ -26,25 +26,34 @@ local core_developers = {
...
+	"Paul Ouellette",
+	"TheTermos",
+	"David CARLIER",
+	"Danila Shutov",
+	"Jean-Patrick Guerrero",
+	"k.h.lai",
```
Comment (SmallJoker) [inline builtin/mainmenu/tab_credits.lua:56]: Combined all areas of work/contributions in a single comment for copy&paste:

```Lua
	"Wuzzy [devtest game, visual corrections]",
	"Zughy [Various fixes]",
	"MoNTE48 [Android improvements]",
	"numzero [Graphics and render fixes]",
	"appgurueu [Various internal fixes]",   -- using GitHub username here
	"Desour [Formspec and vector API changes]", -- commits found using email address
	"HybridDog [Rendering fixes and documentation]",
	"Hugues Ross [Graphics-related improvements]",
	"ClobberXD [Re-binding mouse buttons]",
	"v-rob",-- DUPLICATED: core dev
	"luk3yx [Fixes]",
	"hecks [Audiovisuals, Lua API]",
	"LoneWolfHT [Object crosshair, documentation fixes}",
	"Lejo [Server-related improvements]",
	"EvidenceB [Compass HUD element]", -- company
	"pauloue [GCC warning fix]", -- using GitHub username
	"TheTermos", --?? cannot find any commit
	"David CARLIER [UNIX & Haiku build fixes],
	"Danila Shutov",   --?? cannot find any commit
	"Jean-Patrick Guerrero [Visual fixes]",
	"k.h.lai [Memory leak fixes, documentation]",
```

Comment (appgurueu) [reply builtin/mainmenu/tab_credits.lua:56]: @kilbith @pyrollo how would you like to be attributed?

Comment (EliasFleckenstein03) [reply builtin/mainmenu/tab_credits.lua:56]: `LoneWolfHT [Object crosshair, documentation fixes}` why is this closed by a curly bracket?

Comment (kilbith) [reply builtin/mainmenu/tab_credits.lua]: @SmallJoker how come `model[]` is a “visual fix”? Please fix my attributions.

Comment (sfan5) [reply builtin/mainmenu/tab_credits.lua:56]: > "TheTermos",   -- ?? cannot find any commit

https://github.com/minetest/minetest/commits/master?author=TheTermos

> "Danila Shutov",   -- ?? cannot find any commit

https://github.com/minetest/minetest/commits/master?author=dcbrwn

**Comment by TheTermos** [comment]
I'm not sure about awarding points for changes.
The code is already bloated, and lousy programmers tend to need way more lines to achieve a given goal.
I know that changes work both ways, but, it's much easier to bloat than to optimize.

This system might create an incentive to intentionally write verbose and obscure code just to make it to the list.

File: builtin/mainmenu/tab_credits.lua (hunk 8 above)
Comment (SmallJoker) [reply builtin/mainmenu/tab_credits.lua:56]: Updated accordingly. I used `git log 5.3.0..HEAD --oneline --author="<>"` to filter the commit log but apparently something isn't working correctly there.

**Comment by appgurueu** [comment]
> I'm not sure about awarding points for changes.
> The code is already bloated, and lousy programmers tend to need way more lines to achieve a given goal.
> I know that changes work both ways, but, it's much easier to bloat than to optimize.
> 
> This system might create an incentive to intentionally write verbose and obscure code just to make it to the list.

Unfortunately sort of true; perhaps score should be awarded by core devs?

File: builtin/mainmenu/tab_credits.lua (hunk 8 above)
Comment (Zughy) [reply builtin/mainmenu/tab_credits.lua:56]: Please update mine with `[visual improvements, various fixes]`. Visual ones: #10756 #10152

**Comment by sfan5** [comment]
Given how easy it is to get on the list I doubt that's a problem.
Also when you have a 10 line patch, bloating it up to more than 100 lines (for a mere *one* extra point) would be quite an achievement.

**Comment by rubenwardy** [comment]
sofar should be moved to previous core developers, as they haven't made a contribution since 5.1.0 and I haven't seen them reviewing much either. Ekdohibs perhaps as well, but they've at least made a pull request

I did this with ShadowNinja in the last release

**Comment by TheTermos** [comment]
The point is that the number of changes is quite a dumb metric, just saying.

**Review by rubenwardy — APPROVED** [review]

File: .mailmap [hunk 9]
```diff
@@ -1,33 +1,67 @@
...
+BlockMen <nmuelll@web.de>
+sfan5 <sfan5@live.de>
+DannyDark <the_skeleton_of_a_child@yahoo.co.uk>
+Ilya Pavlov <TTChangeTheWorld@gmail.com>
 sapier <Sapier at GMX dot net> sapier <sapier AT gmx DOT net>
 sapier <Sapier at GMX dot net> sapier <sapier at gmx dot net>
-
+SmallJoker <SmallJoker@users.noreply.github.com> <mk939@ymail.com>
+Loïc Blot <nerzhul@users.noreply.github.com>
+Loïc Blot <nerzhul@users.noreply.github.com> <loic.blot@unix-experience.fr>
```
Comment (nerzhul) [inline .mailmap:34]: @sfan5 why i'm there twice ?

File: .mailmap [hunk 10]
```diff
@@ -1,33 +1,67 @@
...
+numzero <numzer0@yandex.ru> Vitaliy <numzer0@yandex.ru>
+numzero <numzer0@yandex.ru> <silverunicorn2011@yandex.ru>
+Jean-Patrick Guerrero <kilbith@users.noreply.github.com>
+Jean-Patrick Guerrero <kilbith@users.noreply.github.com> <jeanpatrick.guerrero@gmail.com>
+HybridDog <3192173+HybridDog@users.noreply.github.com> <ovvv@web.de>
+srfqi <muhammadrifqipriyosusanto@gmail.com>
+Dániel Juhász <juhdanad@gmail.com>
+rubenwardy <rw@rubenwardy.com>
+rubenwardy <rw@rubenwardy.com> <rubenwardy@gmail.com>
```
Comment (nerzhul) [inline .mailmap:43]: same here

File: util/gather_git_credits.py (hunk 6 above)
Comment (nerzhul) [inline util/gather_git_credits.py:21]: there is no python binding for libgit2 directly ?

**Review by nerzhul — CHANGES_REQUESTED** [review]

File: .mailmap (hunk 9 above)
Comment (sfan5) [reply .mailmap:34]: These lines have different meanings and it's required for it work correctly:
https://git-scm.com/docs/git-check-mailmap#_mapping_authors

File: util/gather_git_credits.py (hunk 6 above)
Comment (sfan5) [reply util/gather_git_credits.py:21]: No idea but for a script ran very rarely this is good enough.

Comment (rubenwardy) [reply util/gather_git_credits.py:21]: I use gitpython in ContentDB, but I suggest just using subprocess for this

**Pull request merged**